### PR TITLE
Refactor CI jobs and test more combinations

### PR DESCRIPTION
**Pull request by imciner2**
Instead of always installing MKL on the CI job, only install it in the ones that use the MKL algebra. This allows for the non-MKL jobs to continue even if there are issues installing it from conda.

### CODE REVIEW

**Comment by imciner2** [comment]
Whoops, I did not want to double the amount of CI jobs we run :grimacing:. Let me look into that.

File: .github/workflows/mkl.yaml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        fail-fast: false
+
+        matrix:
+          # Specify the exact Windows image because the CMake generator in the CI breaks when the image
+          # updates (since Visual Studio is also updated at the same time).
+          os: [ubuntu-latest, macos-latest, windows-2022]
+          python-version: [3.9]
+          long: ['ON', 'OFF']
+
+          # MKL only supports double precision in the iterative solver, so that is the only combination we support as well.
```
Comment (vineetbansal) [inline .github/workflows/mkl.yaml:26]: Perhaps we add `float: ['OFF']` as a single matrix entry to go right next to this comment and use it down below? Just so it remains close to the part of the CI if affects (currently).

**Comment by vineetbansal** [comment]
Looks great - just the one comment, which is not all that important anyway. Thanks Ian!

File: .github/workflows/mkl.yaml (hunk 1 above)
Comment (imciner2) [reply .github/workflows/mkl.yaml:26]: Good idea, that way it better mimics the other CI workflows. Just pushed that change.

**Pull request merged**